### PR TITLE
network: fix bug in inventory advertisement handling

### PR DESCRIPTION
**Pull request by hdevalence**
The RFC https://zebra.zfnd.org/dev/rfcs/0003-inventory-tracking.html described
the use of a `broadcast` channel in place of an `mpsc` channel to get
ring-buffer behavior, keeping a bound on the size of the channel but dropping
old entries when the channel is full.

However, it didn't explicitly describe how this works (the `broadcast` channel
returns a `RecvError::Lagged(u64)` to inform receivers that they lost
messages), so the lag-handling wasn't implemented and I didn't notice in
review.

Instead, the ? operator bubbled the lag error all the way up from
`InventoryRegistry::poll_inventory` through `<PeerSet as Service>::poll_ready`
through various Tower wrappers to users of the peer set.  The error propagation
is bad enough, because it caused client errors that shouldn't have happened,
but there's a worse interaction.

The `Service` contract distinguishes between request errors (from
`Service::call`, scoped to the request) and service errors (from
`Service::poll_ready`, scoped to the service).  The `Service` contract
specifies that once a service returns an error from `poll_ready`, the service
can be assumed to be failed permanently.

I believe (but haven't tested or carefully worked through the details) that
this caused various tower middleware to report the entire peer set service as
permanently failed due to a transient inventory "error" (more of an indicator),
and I suspect that this is the cause of #1003, where all of the sync
component's requests end up failing because the peer set reported that it
failed permanently.  I am able to reproduce #1003 locally before this change
and unable to reproduce it locally after this change, though I have not tested
exhaustively.

### CODE REVIEW

File: zebra-network/src/peer_set/inventory_registry.rs [hunk 1]
```diff
@@ -64,8 +64,33 @@ impl InventoryRegistry {
...
+        // `RecvError::Lagged`. It's crucial that we handle that error here
+        // rather than propagating it through the peer set's Service::poll_ready
+        // implementation, where reporting a failure means reporting a permanent
+        // failure of the peer set.
+        use broadcast::RecvError;
+        while let Poll::Ready(Some(channel_result)) = Pin::new(&mut self.inv_stream).poll_next(cx) {
+            match channel_result {
+                Ok((hash, addr)) => self.register(hash, addr),
+                Err(RecvError::Lagged(count)) => {
+                    tracing::debug!(count, "dropped lagged inventory advertisements");
```
Comment (teor2345) [inline zebra-network/src/peer_set/inventory_registry.rs:89]: Is it worth adding a metric for lagged advertisements?

**Review by yaahc — DISMISSED** [review]

File: zebra-network/src/peer_set/inventory_registry.rs (hunk 1 above)
Comment (hdevalence) [reply zebra-network/src/peer_set/inventory_registry.rs:89]: Yeah, that's a good idea.

**Review by teor2345 — DISMISSED** [review]
Looks good!

I had a question about metrics, but I don't want to block the PR on it.

Do we need to update the RFC as well?
Or does the new comment put the design details in the right place?

File: zebra-network/src/peer_set/inventory_registry.rs (hunk 1 above)
Comment (hdevalence) [reply zebra-network/src/peer_set/inventory_registry.rs:89]: 1800111081710c8332e12a62b5a5d1dd6b77f9dd

**Comment by hdevalence** [comment]
The RFC mentions wanting to use a broadcast channel to get the ring buffer behavior, but it doesn't explicitly describe the lag-handling behavior.  We could leave the description in the comment or make the RFC more explicit, I think either would be fine and it's very little work either way.

**Review by teor2345 — APPROVED** [review]
Thanks, still looks good 👍 

I don't think the exact number of dropped advertisements is important, because it will depend on the number of peers. The number of times we lag is more interesting.

**Review by yaahc — APPROVED** [review]
ship-it!

**Pull request merged**